src/plugins/mer/meroptionswidget.cpp

```diff
@@ -249,60 +250,107 @@ void MerOptionsWidget::onSrcFolderApplyButtonClicked(const QString &newFolder)
     MerSdk *sdk = m_sdks[m_virtualMachine];
 
     if (newFolder == sdk->sharedSrcPath()) {
-        QMessageBox::information(this, tr("Choose a new folder"),
+        QMessageBox::information(this,
+                                 tr("Choose a new folder"),
                                  tr("The given folder (%1) is the current alternative source folder. "
                                     "Please choose another folder if you want to change it.").arg(sdk->sharedSrcPath()));
         return;
     }
 
-    if (!sdk->connection()->isVirtualMachineOff()) {
-        QPointer<QMessageBox> questionBox = new QMessageBox(QMessageBox::Question,
-                tr("Close Virtual Machine"),
-                tr("Close the \"%1\" virtual machine?").arg(m_virtualMachine),
-                QMessageBox::Yes | QMessageBox::No,
-                this);
-        questionBox->setInformativeText(
-                tr("Virtual machine must be closed before the source folder can be changed."));
-        if (questionBox->exec() != QMessageBox::Yes) {
-            // reset the path in the chooser
-            m_ui->sdkDetailsWidget->setSrcFolderChooserPath(sdk->sharedSrcPath());
-            return;
+    const bool vmIsOff = sdk->connection()->isVirtualMachineOff();
+    const bool vmWasClosed = !vmIsOff && showCloseVirtualMachineDialog(tr("Virtual machine must be closed before the source folder can be changed."));
+    bool updateOk = false;
+
+    if (vmIsOff || vmWasClosed) {
+        if (sdk->connection()->lockDown(true)) {
+            updateOk = MerVirtualBoxManager::updateSharedFolder(m_virtualMachine, QLatin1String("src1"), newFolder);
+            sdk->connection()->lockDown(false);
+        }
+
+        if (updateOk) {
+            // Remember to update this value
+            sdk->setSharedSrcPath(newFolder);
+
+            if (vmWasClosed && showStartVirtualMachineDialog(tr("Alternative source folder for %1 changed to %2.").arg(m_virtualMachine).arg(newFolder))) {
+                sdk->connection()->connectTo();
+            }
+        } else {
+            QMessageBox::warning(this,
+                                 tr("Changing the source folder failed!"),
+                                 tr("Unable to change the alternative source folder to %1").arg(newFolder));
         }
     }
 
-    if (!sdk->connection()->lockDown(true)) {
-        QMessageBox::warning(this, tr("Failed"),
-                tr("Alternative source folder not changed"));
-        // reset the path in the chooser
+    if (!updateOk) {
+        // Reset the path in the chooser
         m_ui->sdkDetailsWidget->setSrcFolderChooserPath(sdk->sharedSrcPath());
-        return;
     }
+}
 
-    bool ok = MerVirtualBoxManager::updateSharedFolder(m_virtualMachine,
-            QLatin1String("src1"), newFolder);
+void MerOptionsWidget::onWwwPortApplyButtonClicked(int port)
+{
+    MerSdk *sdk = m_sdks[m_virtualMachine];
```

Comment:
The value of `m_sdks[m_virtualMachine]` is assumed to be non-NULL throughout the class. 

This would appear to be a fairly safe assumption, as `m_virtualMachine` is set to one of the values of the "Virtual Machine" combo box, which in turn is populated with the return value of `m_sdks.keys()`.

In case the last SDK is removed, `m_virtualMachine` is cleared and the SDK details widget is hidden. Because `onWwwPortApplyButtonClicked()` is called via a signal fired by the (now hidden) SDK details widget, `m_sdks[m_virtualMachine]` will not be accessed if there are no SDKs.